File: Abstract Factory/Pizzeria_java/src/main/Pizzeria/Pizza/Pizza.java
package main.Pizzeria.Pizza;

import main.Pizzeria.Pizza.Cheese.ACheese;
import main.Pizzeria.Pizza.Dough.ADough;
import main.Pizzeria.Pizza.Toppings.ATopping;

import java.util.ArrayList;

public final class Pizza {
    private boolean prepared;
    private boolean baked;
    private boolean cut;

    private ADough dough;
    private ACheese cheese;
    private ArrayList<ATopping> toppings;

    public Pizza(ADough dough, ACheese cheese) {
        this.setDough(dough);
        this.setCheese(cheese);
        this.toppings = new ArrayList<>();
        this.setPrepared(false);
        this.setBaked(false);
        this.setCut(false);
    }

    public void prepare() {
        if (!this.isPrepared()) {
            this.setPrepared(true);
            System.out.println("Preparing...");
        }
    }

    public void bake() {
        if (this.isPrepared() && !this.isBaked()) {
            this.setBaked(true);
            System.out.println("Baking...");
        }
    }

    public void cut() {
        if (this.isPrepared() && this.isBaked() && !this.isCut()) {
            this.setCut(true);
            System.out.println("Cutting...");
        }
    }

    public boolean isPrepared() {
        return prepared;
    }

    private void setPrepared(boolean prepared) {
        this.prepared = prepared;
    }

    public boolean isBaked() {
        return baked;
    }

    private void setBaked(boolean baked) {
        this.baked = baked;
    }

    public boolean isCut() {
        return cut;
    }

    private void setCut(boolean cut) {
        this.cut = cut;
    }

    public ADough getDough() {
        return dough;
    }

    private void setDough(ADough dough) {
        if (dough != null)
            this.dough = dough;
        else
            throw new IllegalStateException("Parameter dough must not be null.");
    }

    public ACheese getCheese() {
        return cheese;
    }

    private void setCheese(ACheese cheese) {
        if (cheese != null)
            this.cheese = cheese;
        else
            throw new IllegalStateException("Parameter cheese must not be null.");
    }

    public ArrayList<ATopping> getToppings() {
        return toppings;
    }

    public void addTopping(ATopping topping) {
        if (this.isPrepared() && !this.isBaked())
            if (topping != null)
                if (this.toppings.add(topping))
                    System.out.println("Added: " + topping.getClassName());
            else
                throw new IllegalStateException("Parameter topping must not be null.");
    }

    public void removeTopping(ATopping topping) {
        if (this.isPrepared() && !this.isBaked())
            if (topping != null)
                if (this.toppings.remove(topping))
                    System.out.println("Removed: " + topping.getClassName());
            else
                throw new IllegalStateException("Parameter topping must not be null.");
    }

    public void display() {
        System.out.println("Pizza Ingredients:");
        System.out.println("Dough: " + this.getDough().getClassName());
        System.out.println("Cheese: " + this.getCheese().getClassName());
        System.out.println("Toppings: " + this.getToppings());
    }
}

File: Abstract Factory/Pizzeria_cpp/Pizza.cpp
#include <iostream>

#include "Pizza.h"

Pizza::Pizza(ADough* dough, ACheese* cheese)
{
	this->set_dough(dough);
	this->set_cheese(cheese);
	this->set_prepared(false);
	this->set_baked(false);
	this->set_cut(false);
}

Pizza::~Pizza()
{
	for (const auto topping : this->toppings_)
		delete topping;
	delete this->cheese_;
	delete this->dough_;
}

auto Pizza::prepare() -> void
{
	if (!this->is_prepared()) {
		this->set_prepared(true);
		std::cout << "Preparing...\n";
	}
}

auto Pizza::bake() -> void
{
	if (this->is_prepared() && !this->is_baked()) {
		this->set_baked(true);
		std::cout << "Baking...\n";
	}
}

auto Pizza::cut() -> void
{
	if (this->is_prepared() && this->is_baked() && !this->is_cut()) {
		this->set_cut(true);
		std::cout << "Cutting...\n";
	}
}

auto Pizza::is_prepared() -> bool
{
	return this->is_prepared_;
}

auto Pizza::is_baked() -> bool
{
	return this->is_baked_;
}

auto Pizza::is_cut() -> bool
{
	return this->is_cut_;
}

auto Pizza::get_cheese() -> ACheese*
{
	return this->cheese_;
}

auto Pizza::get_dough() -> ADough*
{
	return this->dough_;
}

auto Pizza::get_toppings() -> std::vector<ATopping*>
{
	return this->toppings_;
}

auto Pizza::add_topping(ATopping* topping) -> bool
{
	if (this->is_prepared() && !this->is_baked()) {
		this->toppings_.push_back(topping);
		std::cout << "Added: " << topping->get_class_name() << "\n";
		return true;
	}
	return false;
}

auto Pizza::remove_topping(ATopping* topping) -> bool
{
	if (this->is_prepared() && !this->is_baked())
	{
		for (const auto _topping : this->toppings_)
		{
			if (_topping == topping) {
				delete _topping;
				return true;
			}
		}
	}
	return false;
}

auto Pizza::set_prepared(bool is_prepared) -> void
{
	this->is_prepared_ = is_prepared;
}

auto Pizza::set_baked(bool is_baked) -> void
{
	this->is_baked_ = is_baked;
}

auto Pizza::set_cut(bool is_cut) -> void
{
	this->is_cut_ = is_cut;
}

auto Pizza::set_cheese(ACheese* cheese) -> void
{
	this->cheese_ = cheese;
}

auto Pizza::set_dough(ADough* dough) -> void
{
	this->dough_ = dough;
}
